code/obj/displaycase.dm

```diff
@@ -382,3 +342,89 @@
 
 		//DEBUG_MESSAGE("[src.name]'s quality_counter: [quality_counter]")
 		return
+
+/datum/action/bar/icon/captaingun_assembly
+	id = "captaingun_assembly"
+	interrupt_flags = INTERRUPT_MOVE | INTERRUPT_ACT | INTERRUPT_ATTACKED | INTERRUPT_STUNNED
+	icon = 'icons/ui/actions.dmi'
+	icon_state = "working"
+	duration = 3.5 SECONDS
+
+	var/obj/item/captaingun/gun
+	var/obj/item/stage_item
+
+	New(var/obj/item/captaingun/gun/O, var/obj/item/I)
+		..()
+		if(O)
+			src.gun = O
+		if(I)
+			src.stage_item = I
+			src.icon = I.icon
+			src.icon_state = I.icon_state
+
+	onUpdate()
+		..()
+		if(QDELETED(src.gu)n || QDELETED(src.stage_item) || BOUNDS_DIST(owner, gun) > 0 || BOUNDS_DIST(owner, src.stage_item) > 0)
+			interrupt(INTERRUPT_ALWAYS)
+			return
+		var/mob/source = owner
+		if(istype(source) && stage_item != source.equipped())
+			interrupt(INTERRUPT_ALWAYS)
+
+	onStart()
+		..()
+		switch(gun.repair_stage)
+			if(1)
+				boutput(owner, "<span class='notice'>You begin to rewire the gun's circuit board...</span>")
+			if(2)
+				boutput(owner, "<span class='notice'>You begin to install the coil...</span>")
+			if(3)
+				boutput(owner, "<span class='notice'>You begin to solder the coil into place...</span>")
+			if(4)
+				boutput(owner, "<span class='notice'>You begin to install the lens...</span>")
+			if(6)
+				boutput(owner, "<span class='notice'>You begin to install the power cell...</span>")
+
+	onEnd()
+		..()
+		var/mob/user = owner
```

Comment:
I don't see thee point of this var?

Reply:
owner is not a mob and subsequent calls need a mob to work (like u_equip)